### ISSUE TITLE
[OSX]corerun CoreClrHelloWorld fails in stackwalktypes.h

### ISSUE THREAD

**Issue by cnblogs-dudu** [dotnet/coreclr#164]
Follow [Building and Running CoreCLR on Linux](https://github.com/dotnet/coreclr/wiki/Building-and-Running-CoreCLR-on-Linux):
* build CoreClrHelloWorld on Mac OS X
* Transfer CoreClrHelloWorld to Linux Ubuntu
* copy "corerun, libcoreclr.so, mscorlib.dll" on Linux into CoreClrHelloWorld

![screen shot 2015-02-09 at 16 22 58](https://cloud.githubusercontent.com/assets/3196961/6102866/064d622c-b078-11e4-9190-2ec8770762c1.png)
Then execute `/data/git/CoreClrHelloWorld# runtime_linux/corerun app/HelloWorld.exe`
Run successfully on Linux.

Back to OSX, copy "corerun, libcoreclr.dylib, mscorlib.dll" on OSX into CoreClrHelloWorld

![screen shot 2015-02-09 at 16 33 45](https://cloud.githubusercontent.com/assets/3196961/6102982/7f4e6d3c-b079-11e4-91fb-88ebe79575b8.png)
Then execute `CoreClrHelloWorld dudu$ runtime_mac/corerun app/HelloWorld.exe`.
Run fails:
```
Hello, Windows...

  Assert failure (unable to format)
/git/dotnet/coreclr/src/vm/stackwalktypes.h
SPOffset >= pUnwindInfo->RSPOffsetFromUnwindInfo
**** MessageBox invoked, title 'Assert failure (unable to format)' ****
  SPOffset >= pUnwindInfo->RSPOffsetFromUnwindInfo
********

Assert failure (unable to format)
/git/dotnet/coreclr/src/vm/stackwalktypes.h
FitsIn<DWORD>(pUnwindInfo->RBPOffset + (SPOffset - pUnwindInfo->RSPOffsetFromUnwindInfo))
**** MessageBox invoked, title 'Assert failure (unable to format)' ****
  FitsIn<DWORD>(pUnwindInfo->RBPOffset + (SPOffset - pUnwindInfo->RSPOffsetFromUnwindInfo))
********

libc++abi.dylib: terminating with uncaught exception of type PAL_SEHException
Abort trap: 6
```
(View `_ASSERTE(SPOffset >= pUnwindInfo->RSPOffsetFromUnwindInfo);` in stackwalktypes.h: https://github.com/dotnet/coreclr/blob/master/src/vm/stackwalktypes.h#L154)

When comment all "Console.BackgroundColor, Console.ForegroundColor, Console.ResetColor()" code in [HelloWorld.cs](https://github.com/cnblogs-dudu/CoreClrHelloWorld/commit/2bfecd6e43329916096226d5fb2f69645612e0d5), CoreClrHelloWorld  run  successfully!

![screen shot 2015-02-09 at 16 51 24](https://cloud.githubusercontent.com/assets/3196961/6103175/ff16fadc-b07b-11e4-9ded-36afecc3f931.png)

If you want to try it, clone from the [CoreClrHelloWorld](https://github.com/cnblogs-dudu/CoreClrHelloWorld) repository on GitHub.

**Comment by ellismg** [comment]
My hunch is that the code in System.Console.dll which reads terminfo files doesn't work correctly on OSX.  If you're up for it, look at ConsolePal.Unix.cs in the CoreFX repository, to see what we are trying to do.  You might also try setting COMPlus_DumpJittedMethods=1 in your environment to get a better idea of where this is crashing.  I expect the runtime is trying to throw an exception, which causes stack walking which may not be 100% on OSX yet.

**Comment by cnblogs-dudu** [comment]
@ellismg 
"COMPlus_DumpJittedMethods=1" logs: https://gist.github.com/cnblogs-dudu/ace5d9039716860f6a8c

**Comment by cnblogs-dudu** [comment]
The 'SPOffset >= pUnwindInfo->RSPOffsetFromUnwindInfo' failture is produced with `mcs HelloWorld.cs` command. This is not really problem.
When compiled with the following command:
```
mcs /nostdlib /noconfig /r:../packages/System.Console.4.0.0-beta-22512/lib/contract/System.Console.dll /r:../packages/System.Runtime.4.0.20-beta-22512/lib/contract/System.Runtime.dll HelloWorld.cs
```
There is following exception:
```
Compiling  126 DomainBoundILStubClass::IL_STUB_PInvoke, IL size = 39, hsh=0x449ffd05
{0x7fff75a7a300-0x10fbef6b0} ASSERT [MISC   ] 
at /git/dotnet/coreclr/src/pal/src/misc/fmtmessage.cpp.239: 
Unable to continue due to missing library.
libc++abi.dylib: terminating with uncaught exception of type PAL_SEHException
Abort trap: 6
```

**Comment by kangaroo** [comment]
The stack pointer offset is a real problem, but I'm not going to work on it until we reach an agreement on https://github.com/dotnet/coreclr/issues/177

For the IL_STUB_PInvoke issue using the contract assemblies, the issue is in corefx.  System.Console currently uses open64 to parse terminfo's which doesn't exist on OSX.

**Comment by ellismg** [comment]
@stephentoub FYI, for the Console issue.

**Comment by kangaroo** [comment]
Console issue filed @ https://github.com/dotnet/corefx/issues/715

**Comment by kangaroo** [comment]
@cnblogs-dudu If you really want to make it work you can build System.Console with my patch in https://github.com/dotnet/corefx/issues/715  Alternately I posted binaries here*:

https://www.dropbox.com/s/samo85xvn8c2gnw/coreclr-bcl-osx.zip?dl=0

*NOTE: I was too lazy to patch the corlib for this build so you need to ln -s libcoreclr.dylib libcoreclr.so

**Comment by cnblogs-dudu** [comment]
@kangaroo Thanks. Got it. I will try it.

**Comment by cnblogs-dudu** [comment]
@kangaroo Great! It works. Compile HelloWorld.cs with your binaries and corerun with your previous mscorlib.dll successfully.
```
Hello, Mac...

                                   ,++
                               @@@@@@+
                             @@@@@@@@
                           @@@@@@@@@:
                          @@@@@@@@@.
                         :@@@@@@@@
                         @@@@@@;
                         `
         ,@@@@@@@@@@@:        @@@@@@@@@@@@@:
      :@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@:
    :@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
   @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
#@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
:@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@'
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@`
  @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
   @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
     @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
      @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
       @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
         @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
          :@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@;
             @@@@@@@;           ;@@@@@@@

Press ENTER to exit ...
```

**Comment by kangaroo** [comment]
Colors are broken tho.  Happy to review PRs.

**Comment by kangaroo** [comment]
Can we close this issue now?  The root of the stack unwinding is covered in https://github.com/dotnet/coreclr/issues/177 which I will for into an OS X task once available.

**Comment by cnblogs-dudu** [comment]
@kangaroo If fix this issue - [dotnet/corefx#723](https://github.com/dotnet/corefx/issues/723), color will emerge.
